crypto/cipher_extra/e_aes_cbc_hmac_sha1.c

```diff
@@ -0,0 +1,362 @@
+                        const AES_KEY *key, unsigned char iv[16], SHA_CTX *ctx,
+                        const void *in0);
+
+static int aesni_cbc_hmac_sha1_init_key(EVP_CIPHER_CTX *ctx,
+                                        const unsigned char *inkey,
+                                        const unsigned char *iv, int enc) {
+  EVP_AES_HMAC_SHA1 *key = data(ctx);
+  int ret;
+
+  int key_bits = EVP_CIPHER_CTX_key_length(ctx) * 8;
+  if (enc) {
+    ret = aes_hw_set_encrypt_key(inkey, key_bits, &key->ks);
+  } else {
+    ret = aes_hw_set_decrypt_key(inkey, key_bits, &key->ks);
+  }
+
+  SHA1_Init(&key->head);
+  key->tail = key->head;
+  key->md = key->head;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  return ret < 0 ? 0 : 1;
+}
+
+static int aesni_cbc_hmac_sha1_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                                      const unsigned char *in, size_t len) {
+  EVP_AES_HMAC_SHA1 *key = data(ctx);
+  size_t plen = key->payload_length, iv = 0;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  if (len % AES_BLOCK_SIZE) {
+    OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_DATA_NOT_MULTIPLE_OF_BLOCK_LENGTH);
+    return 0;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    // NOTE: Difference between openssl and aws-lc:
+    // In encrypt case, |plen| is set in the call |EVP_CIPHER_CTX_ctrl| with
+    // |EVP_CTRL_AEAD_TLS1_AAD| operation.
+    // When |plen == NO_PAYLOAD_LENGTH|, it means the call did not happen.
+    // In this case, aws-lc returns error(0) but openssl supports that with
+    // below explanation.
+    // https://mta.openssl.org/pipermail/openssl-users/2019-November/011458.html
+    // -- These stitched ciphers are specifically targeted at use by libssl
+    //    and are designed for use in SSL/TLS only.
+    if (plen == NO_PAYLOAD_LENGTH) {
+      // |EVP_CIPHER_CTX_ctrl| with |EVP_CTRL_AEAD_TLS1_AAD| operation is not
+      // performed.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_CTRL_OPERATION_NOT_PERFORMED);
+      return 0;
+    }
+    if (len !=
+        ((plen + SHA_DIGEST_LENGTH + AES_BLOCK_SIZE) & -AES_BLOCK_SIZE)) {
+      // The input should have space for eiv + plaintext + digest + padding.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_UNSUPPORTED_INPUT_SIZE);
+      return 0;
+    } else if (key->aux.tls_ver >= TLS1_1_VERSION) {
+      iv = AES_BLOCK_SIZE;
+    }
+
+    size_t aes_off = 0;
+    size_t sha_off = SHA_CBLOCK - key->md.num;
+    size_t blocks;
+
+    // TODO: check CPU cap like e_aes_cbc_hmac_sha256.c?
+    if (plen > (sha_off + iv) &&
+        (blocks = (plen - (sha_off + iv)) / SHA_CBLOCK)) {
+      SHA1_Update(&key->md, in + iv, sha_off);
+
+      aesni_cbc_sha1_enc(in, out, blocks, &key->ks,
+                         EVP_CIPHER_CTX_iv_noconst(ctx), &key->md,
+                         in + iv + sha_off);
+      // Update the offset to record and skip the part processed
+      // (encrypted and hashed) by |aesni_cbc_sha1_enc|.
+      blocks *= SHA_CBLOCK;
```

Comment:
I think that we really need an explanation of how the encryption (and decryption below) works. Just looking at the code it's very difficult to deduct what's happening, what `aes_off, sha_off, blocks` represent, what's `key->md.Nh/l`, how is `block` computed and updated, the same for `aes_off, sha_off`, ... I'm willing to help out with writing a good explanation so feel free to ping me offline.